discoverd2/server/state.go

```diff
@@ -0,0 +1,314 @@
+package server
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+type EventKind uint
+
+const (
+	EventKindUp EventKind = 1 << iota
+	EventKindUpdate
+	EventKindDown
+	EventKindLeader
+)
+
+func (k EventKind) String() string {
+	switch k {
+	case EventKindUp:
+		return "up"
+	case EventKindUpdate:
+		return "update"
+	case EventKindDown:
+		return "down"
+	case EventKindLeader:
+		return "leader"
+	default:
+		return "unknown"
+	}
+}
+
+type Event struct {
+	Service string
+	Kind    EventKind
+	*Instance
+}
+
+func eventKindUpdate(existing bool) EventKind {
+	if existing {
+		return EventKindUpdate
+	}
+	return EventKindUp
+}
+
+type Instance struct {
```

Comment:
This will likely be moved to the client package, named `discoverd`. `ServiceInstance` felt like a mouthful, but if you think it's better I'll gsub.